contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2867,6 +2879,377 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
+						* "YYYY MM DD", "DD{/.-}MM{/.-}YYYY", "DD{/.-}YYYY{/.-}DD"
+						* then these dates shouldn't be supported.
+						* Supported date formats are `Mon yyyy dd`, `Mon dd yyyy`,
+						* `mm{-/.}dd{-/.}yyyy`, `yyyy{-/.}mm{-/.}dd`
+						*/
+						for (int k = 0 ; k < 3; k++)
+						{
+							temp_field = strtok(temp_field, different_date_formats[k]);
+							if (pg_strcasecmp(field[i], temp_field) != 0)
+								break;
+						}
+						/*
+						* If text month is present in start of field like
+						* `Mon{/.-}yyyy{/.-}dd` or `Mon{/.-}dd{/.-}yyyy then
+						* an error should be thrown
+						*/
+						if (isTextMonthPresent(temp_field))
+							TIME_IN_ERROR();
+					}
+					break;
+
+				case DTK_TIME:
+					/*
+					 * If the input is of format '<DATE>T<TIME>, then the
+					 * <TIME> should be of format hh:mm:ss[.nnn]
+					 * where hh should be of 2 digit.
+					 */
+					if (i-1 == 1 && field[1] && pg_strcasecmp(field[1], "t") == 0)
+						pattern = "^([0-1][0-9]|2[0-3])(:[0-5]?[0-9]:[0-5]?[0-9]?.[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9])?$";
+					else
+						/* check if tsql time format in time type follows hh:mm[:ss][.nnn] format */
+						pattern = "^([0-1]?[0-9]|2[0-3])(:[0-5]?[0-9])(:[0-5]?[0-9]?.[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9]?[0-9])?$";
+					if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+						ereport(ERROR,
+									(errcode(ERRCODE_INTERNAL_ERROR),
+									errmsg("time format regex failed")));
+					if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+						TIME_IN_ERROR();
+					regfree(&time_regex);
+					break;
+
+				case DTK_TZ:
+					/*
+					 * If the input contains timezone then the format should be "[+-]ZZ:ZZ" and the previous field should be time.
+					 */
+					if(i-1 < 0 || (i-1 >= 0 && ftype[i-1] != DTK_TIME))
+						TIME_IN_ERROR();
+					pattern = "([0-9]?[0-9]:[0-9]?[0-9])$";
+					if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+						ereport(ERROR,
+									(errcode(ERRCODE_INTERNAL_ERROR),
+									errmsg("time format regex failed")));
+					if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+						TIME_IN_ERROR();
+					regfree(&time_regex);
+
+				default:
+					break;
+			}
+			/* Free the temp_field which stores the field[i] value */
+			pfree(temp_field);
+		}
+		switch (nf)
+		{
+			case 1:
+				/*
+				 * If only date is specified then add an default
+				 * time of 00:00:00
+				 */
+				if (ftype[0] == DTK_DATE)
+				{
+					ftype[1] = DTK_TIME;
+					field[1] = "00:00:00";
+					nf = nf + 1;
+ 				}
+				break;
+
+			case 3:
+				/*
+				 * If the given input is of valid format "YYYY-MM-DDThh:mm:ss"
+				 * then conver it to "YYYY-MM-DD hh:mm:ss" by ignoring 'T'.
+				 * Invalid formats are "YYYY-MM-DDthh:mm:ss", "YYYY-MM-DDT hh:mm:ss"
+				 */
+				if (ftype[1] == DTK_STRING && pg_strcasecmp(field[1], "t") == 0 && ftype[2] == DTK_TIME)
+				{
+					/* If the input contains "t" then throw error */
```

Comment:
this is confusing, why would you throw error if input contains "t"? or is it that a T in input is converted to "t" in field[1]? also why are we only checking for "t", PG support both t/T and SQL server only T

Reply:
While parsing all the [A-Z] will be converted to [a-z], so the fields will always contain [a-z]. But for the input format "<DATE>T<TIME>" the "T" should always be in caps. So we are checking whether the original str contains "T" or "t", if "t" is present then we are throwing the error. All of this done only after checking that the str is of forma "<DATE>T<TIME>".